### ISSUE TITLE
System Calls change?

### ISSUE THREAD

**Issue by LaifsV1** [runtimeverification/k#830]
In revision eb4c881, system calls for Z3 always return either "sat/n" or "unsat/n".
In revision b1e46e5, system calls behave differently, and they sometimes return an empty string when manually running the call returns either "sat/n" or "unsat/n".
Did something change with the way System Call Results are handled?

I was trying to call Z3 this way:
```
  rule CheckSAT(P) => SystemSAT(#system("echo \"" +String LibToString(P) +String "(check-sat)\" | z3 -in"))
  rule SystemSAT(#systemResult ( 0, "sat\n" ,   _)) => true
  rule SystemSAT(#systemResult ( 0, "unsat\n" , _)) => false
```
edit: I'm running the platform independent binaries

**Comment by dwightguth** [comment]
@nishantjr can you shed any light on this?

**Comment by LaifsV1** [comment]
To give some more information, I did a binary search through the commits to see where my semantics break, and all releases up to July 09 work correctly. That is, commit `0d23638`. After that, the first commit with platform independent binaries that behaves differently is `01ba514c0`, released 30 Jul.

**Comment by nishantjr** [comment]
Which backend are you using? I've been calling to CVC4 extensively using the OCAML backend, and not observed this behaviour. I wonder if its due to longer running calls... ?

**Comment by LaifsV1** [comment]
Ah, I'm using the Java backend, since I'm using the substitution module.

**Comment by nishantjr** [comment]
When trying to upstream the K smt library, I had noticed test failures where it appears that Z3 does not read the entire input, and thus prints the wrong output. I wonder if it's a related bug. I wasn't able to get far with that

**Comment by LaifsV1** [comment]
When manually inputting the formula, it does work though. It's only when called from K. Stderr is also empty, and only a `\n` is printed.

**Comment by LaifsV1** [comment]
To give an idea, the formulas are very small. I'm checking path conditions which are maybe 10 clauses. So I don't expect Z3 to be taking too long on them.

**Comment by nishantjr** [comment]
Do the queries deterministically fail, or are they flaky and fail sometimes?

**Comment by LaifsV1** [comment]
So far it appears to be deterministic. The first call is successful and returns "sat\n". The next call appears to be returning an empty string followed by a new line (I'm printing out the result in a log file) and the semantics gets stuck. The function appears to not return true or false correctly afterwards.

i.e. my rules return true if sat, false if unsat, and instead of reaching a final configuration, it returns a freezer state. Logging the call result in a file  I usually see a single successful call, e.g. "sat\n", followed by another new line. It is possible I'm not logging the call result correctly though. I need to test this more carefully.

Testing with the early July release of K, I wasn't able to reproduce this behaviour as the calls always succeeds and the semantics didn't get stuck. I could be mistaken about the reason, but something appears to have changed the behaviour between the releases.

**Comment by LaifsV1** [comment]
An update:
On Ubuntu 16.04, 18.10 and 19.04, the old K version always returns sat or unsat correctly.
On Ubuntu 16.04 VM running on Windows 10, same as running directly on Ubuntu.
On Ubuntu 16.04 WSL on Windows 10, it breaks.

I couldn't get the Windows 10 Ubuntu to work. Not even the older version of K does what I expected. 

After fixing my file logging, it seems that several successful calls are performed. All calls logged had a successful sat or unsat result. But then, it gets stuck. Since the call isn't even logged, it's possible the call isn't even made. However, the semantics gets stuck and returns a freezer state where a call is expected to be made. If I replace the call with true, the semantics doesn't get stuck.

**Comment by LaifsV1** [comment]
After some time, I've tested this again, and I am still stumped as to why system calls won't work reliably.
When testing, this:
```
  syntax KResult ::= Bool  
  syntax Bool ::= CheckSAT(SMTLib)  [function]
  syntax Bool ::= SystemSAT(KItem)  [function,strict(1)]

  rule SystemSAT(#systemResult ( _, "sat\n" ,   _)) => true
```
gets stuck and does not return true.

However, this:
```
  rule SystemSAT(#systemResult ( _, _,   _)) => true
```
Does not get stuck. I've logged the result of the call, and it is indeed `sat\n`.

I then attempted to log the result being passed by K:
```
  rule SystemSAT(#systemResult ( A, B, C)) => SystemLOG(#system("echo '" +String B +String "' > log.txt"))
  rule SystemLOG(#systemResult ( _, _, _)) => true
```
This successfully logged `sat\n` in the file, and the semantics did not get stuck.
However, when changing the `>` to a `>>`, the semantics gets stuck.
The following rule gets stuck without logging anything to file:
```
  rule SystemSAT(#systemResult ( A, B, C)) => SystemLOG(#system("echo '" +String B +String "' >> log.txt"))
  rule SystemLOG(#systemResult ( _, _, _)) => true
```
I can't explain this. Either I'm doing something wrong, or system calls are not returning correctly.

**Comment by nishantjr** [comment]
Thanks for documenting this here. I'd recommend moving away from the Java backend.
The LLVM backend is much closer to feature parity now (for concrete execution).

**Comment by LaifsV1** [comment]
Thanks for the reply! 
Is it just --backend llvm?
I get an error `Compiler: No fresh generator defined for sort Optional[KVar]`, what does this mean?

**Comment by nishantjr** [comment]
Would you mind asking this question on Stackoverflow? It would add to the publicly available, easily googleable K knowledge base. Just tag your question with `kframework`.

Thanks!

**Comment by LaifsV1** [comment]
To document an update on this, I tried the LLVM backend and couldn't get it to compile. 

I decided to try the Java backend again with the lastest release and the system calls are working again. I have found that the system calls were succeeding but not being pattern matched correctly. Particularly the underscores `_` are not matched correctly here. I suspect it might have been a problem with scope of the underscore.

i.e.
```
rule SystemSAT(#systemResult ( _, "sat\n" ,   _)) => true
```
needs to be converted to:
```
rule SystemSAT(#systemResult ( 0, "sat\n" ,  "")) => true
```
for the system call return to be matched correctly.

Trying the version with the `_` in Ubuntu seems to work up to 01ba5140 (Jul 30 2019) on different machines, but not in other operating systems. In the newest release, the version without `_` works in MacOS.

Is the behaviour of `_`  working as intended? any explanation why `_` doesn't match the return?
